File: julia/2021/day20.jl
using Test

function process(algo, image, default = 0)
    default = algo[1] == '.' ? 0 : default

    new_image = Dict()
    min_i = 1000
    max_i = -1000
    min_j = 1000
    max_j = -1000

    for (i, j) in keys(image)
        min_i = min(i, min_i)
        min_j = min(j, min_j)
        max_i = max(i, max_i)
        max_j = max(j, max_j)
    end

    for i = min_i-1:max_i+1
        for j = min_j-1:max_j+1
            n::UInt16 = 0
            for a = i-1:i+1
                for b = j-1:j+1
                    n <<= 1
                    n |= get(image, (a, b), default)
                end
            end

            new_image[(i, j)] = algo[n+1] == '#' ? 1 : 0
        end
    end

    return new_image
end

function part_one(input, times = 2)
    (algo, image) = split(input, "\n\n")

    image = split(image, "\n")

    image_dict = Dict()

    for i = 1:length(image)
        for j = 1:length(image[1])
            image_dict[(i, j)] = image[i][j] == '#' ? 1 : 0
        end
    end

    for a = 0:times-1
        image_dict = process(algo, image_dict, a % 2)
    end

    return reduce(+, values(image_dict))
end

function part_two(input)
    return part_one(input, 50)
end

function test()
    example = """..#.#..#####.#.#.#.###.##.....###.##.#..###.####..#####..#....#..#..##..###..######.###...####..#..#####..##..#.#####...##.#.#..#.##..#.#......#.###.######.###.####...#.##.##..#..#..#####.....#.#....###..#.##......#.....#..#..#..##..#...##.######.####.####.#.#...#.......#..#.#.#...####.##.#......#..#...##.#.##..#...##.#.##..###.#......#.#.......#.#.#.####.###.##...#.....####.#..#..#.##.#....##..#.####....##...##..#...#......#.#.......#.......##..####..#...#.#.#...##..#.#..###..#####........#..####......#..#

#..#.
#....
##..#
..#..
..###"""

    @test part_one(example) == 35
    @test part_two(example) == 3351
end

File: ruby/2023/day20.rb
require './commons/assert'

def initial_configuration(input)
  modules = input.split("\n").each_with_object({}) do |l, m|
    name, targets = l.split(' -> ')
    if name[0] == '%'
      # Flip Flop, state = HIGH (true) or LOW (false)
      m[name[1..]] = ['%', name[1..], targets.split(', '), false]
    elsif name[0] == '&'
      # Conjunction, state = { input name => pulse }
      m[name[1..]] = ['&', name[1..], targets.split(', '), {}]
    else
      m[name] = ['B', name, targets.split(', ')]
    end
  end

  # Need this to freeze the keys, as we might add new (output only) modules
  modules.each_key.to_a.each do |k|
    modules[k][2].each do |t|
      modules[t] ||= [t, t, []]
      modules[t][3][k] = false if modules[t][0] == '&'
    end
  end

  modules
end

def process_pulse(mod, pulse)
  case mod[0]
  when '%'
    return [] if pulse[1] == true

    mod[3] = !mod[3]
    mod[2].map { [_1, [mod[1], mod[3]]] }
  when '&'
    mod[3][pulse[0]] = pulse[1]
    mod[2].map { [_1, [mod[1], !mod[3].each_value.all?]] }
  when 'B'
    mod[2].map { [_1, [mod[1], pulse[1]]] }
  else
    []
  end
end

def part_one(input)
  modules = initial_configuration(input)

  highs = 0
  lows = 0
  1000.times do
    queue = [['broadcaster', ['button', false]]]
    until queue.empty?
      target, pulse = queue.shift
      queue.concat(process_pulse(modules[target], pulse))
      highs += 1 if pulse[1]
      lows += 1 unless pulse[1]
    end
  end
  highs * lows
end

def part_two(input)
  modules = initial_configuration(input)

  # rx will receive its low pulse when &jz receives all highs (4 inputs) -> Manual inspection of input file
  highs_in_jz = []
  presses = 0
  loop do
    presses += 1
    queue = [['broadcaster', ['button', false]]]
    until queue.empty?
      target, pulse = queue.shift
      queue.concat(process_pulse(modules[target], pulse))

      # We save when jz has a high pulse, then get the lcm of the four inputs to check when all 4 will cycle together
      if target == 'jz' && pulse[1]
        highs_in_jz << presses
        # p [target, pulse, presses]
      end

      return highs_in_jz.reduce(1, :lcm) if highs_in_jz.size == 4
    end
  end
end

def test
  example = %(broadcaster -> a, b, c
%a -> b
%b -> c
%c -> inv
&inv -> a)

  example2 = %(broadcaster -> a
%a -> inv, con
&inv -> b
%b -> con
&con -> output)

  assert part_one(example), 32_000_000
  assert part_one(example2), 11_687_500
  0
end
